src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreBuildHandler.cs

```diff
@@ -117,6 +120,17 @@ public void Dispose()
 
         public void UpdateSolution_QueryDelayBuildAction(uint dwAction, out IVsTask pDelayTask)
         {
+            if (!_isMEFInitialized)
+            {
+                ThreadHelper.JoinableTaskFactory.Run(async () =>
```

Comment:
Yes, `UpdateSolution_QueryDelayBuildAction` is guaranteed to execute before `RestoreAsync` since UpdateSolution is the event triggered by SBM for any solution event like build, rebuild, etc... `RestoreAsync` is public only for the testing purpose.